sources/engine/Stride.Engine/Profiling/GameProfilingSystem.cs

```diff
@@ -99,131 +66,49 @@ public GameProfilingSystem(IServiceRegistry registry) : base(registry)
+        private async Task UpdateFpsAsync(CancellationToken cancellationToken)
+        {
+            while (!backgroundUpdateCancellationSource.IsCancellationRequested)
             {
-                AppendEvent(profilingResults[((int)CurrentResultPage - 1) * elementsPerPage + i], elapsedFrames, containsMarks);
-            }
-            profilingResults.Clear();
-
-            if (numberOfPages > 1)
-                profilersStringBuilder.AppendFormat("PAGE {0} OF {1}", CurrentResultPage, numberOfPages);
+                //In Fps filtering mode wait until it is time to update the output.
+                await Task.Delay((int)RefreshTime, cancellationToken).ConfigureAwait(false);
 
-            const float mb = 1 << 20;
-
-            gpuInfoStringBuilder.Clear();
-            gpuInfoStringBuilder.AppendFormat("Drawn triangles: {0:0.0}k, Draw calls: {1}, Buffer memory: {2:0.00}[MB], Texture memory: {3:0.00}[MB]", trianglesCount / 1000f, drawCallsCount, GraphicsDevice.BuffersMemory / mb, GraphicsDevice.TextureMemory / mb);
-
-            gpuGeneralInfoStringBuilder.Clear();
-            //Note: renderTargetSize gets set in Draw(), without synchronization, so might be temporarily incorrect.
-            gpuGeneralInfoStringBuilder.AppendFormat("Device: {0}, Platform: {1}, Profile: {2}, Resolution: {3}", GraphicsDevice.Adapter.Description, GraphicsDevice.Platform, GraphicsDevice.ShaderProfile, renderTargetSize);
-
-            fpsStatStringBuilder.Clear();
-            fpsStatStringBuilder.AppendFormat("Displaying: {0}, Frame: {1}, Update: {2:0.00}ms, Draw: {3:0.00}ms, FPS: {4:0.00}", FilteringMode, Game.DrawTime.FrameCount, Game.UpdateTime.TimePerFrame.TotalMilliseconds, Game.DrawTime.TimePerFrame.TotalMilliseconds, Game.DrawTime.FramePerSecond);
-
-            lock (stringLock)
-            {
-                gcCollectionsString = gcCollectionsStringBuilder.ToString();
-                gcMemoryString = gcMemoryStringBuilder.ToString();
-                profilersString = profilersStringBuilder.ToString();
-                fpsStatString = fpsStatStringBuilder.ToString();
-                gpuInfoString = gpuInfoStringBuilder.ToString();
-                gpuGeneralInfoString = gpuGeneralInfoStringBuilder.ToString();
+                UpdateStringsAndProfilers(containsMarks: false);
             }
         }
 
-        private async Task ReadEventsAsync()
+        private async Task UpdateWithProfilerEventsAsync(CancellationToken cancellationToken)
         {
             var containsMarks = false;
+            Task delayTask = Task.Delay((int)RefreshTime, cancellationToken);
 
-            //TODO: Untangle this a bit. Currently fps display (FilteringMode == GameProfilingResults.Fps)
-            //      depends on the timer/update logic, but it does not actually need the profiling events.
-            
-            while (Enabled)
+            while (await profilerChannel.WaitToReadAsync(cancellationToken))
```

Comment:
I don't think we need to support cancellation here. If we're running this task and request cancellation, we've already called Profiler.Unsubscribe(), which will complete the channel, ending this loop. (And we must call Unsubscribe() to let the Profiler know that the channel should be removed). And if we don't support cancellation the call will also no longer allocate.